src/osd/OSD.cc

```diff
@@ -7865,6 +8085,73 @@ void OSD::consume_map()
     assert(newly_split.empty());
   }
 
+  // prime merges
+  set<spg_t> merge_pgs;
+  // build a set of pgids that are sources or targets for any other pg
+  // that we store that will merge.
+  for (auto& shard : shards) {
+    shard->identify_merges(osdmap, &merge_pgs);
+  }
+  // FIXME, maybe: We could race against an incoming peering message
+  // that instantiates a merge PG after identify_merges() below and
+  // never set up its peer to complete the merge.  An OSD restart
+  // would clear it up.  This is a hard race to resolve,
+  // extraordinarily rare (we only merge PGs that are stable and
+  // clean, so it'd have to be an imported PG to an OSD with a
+  // slightly stale OSDMap...), so I'm ignoring it for now.  We plan to
+  // replace all of this with a seastar-based code soon anyway.
```

Comment:
is there any way to crash and die in a safe way in case the race is triggered? or is that where we'll end up going in that event?

Reply:
Could be.  The failure mode here is that the PG may end up with a queue waiter item that stalls future events.. not a crash. And 50% of the time that will be the destroyed "source" pg.  I'll think about whether it can be mopped up, though.  I just worry that the code dealing with that case won't ever get tested so it's just as likely to break as to make things better.